posts/parent-class/index.qmd

```diff
@@ -0,0 +1,50 @@
+---
+title: "Choosing the Right Parent for R Object Classes"
+author:
+  - name: "Hugo Gruson"
+    orcid: "0000-0002-4094-1476"
+date: "2024-06-24"
+categories: [R, R package, object-oriented programming, S3, interoperability]
+---
+
+[I have recently published](https://hugogruson.fr/posts/compa-tibble/) a [series of blog posts](https://epiverse-trace.github.io/posts/s3-generic/) on the reasons why one may want to start using object-oriented programming (and more specifically R S3 classes) to improve interoperability with other tools from the ecosystem.
+
+But there are still questions I have not addressed directly, even if they may have been implicitly included sometimes: what makes a good object class? What good practices in class & function design can improve interoperability?
+
+As you can expect from these questions, this post may be slightly more subjective than the previous ones on this topic.
+
+I will argue that it is often a good strategy to inherit from existing standards classes, and to leverage this inheritance relationship as much as possible.
+
+## Inherit from standard classes
+
+A peculiarity of R is the availability and centrality of `data.frame`s in the base language, whereas you need extra libraries for a similar functionality in most other languages (e.g., pandas in Python).
+
+`data.frame` is one of the first "complex" (in the sense of [non-atomic](https://adv-r.hadley.nz/vectors-chap.html#atomic-vectors)) object most R learners will be exposed to and will develop a familiarity with. A good way to leverage this familiarity is to make your subclass a thin wrapper around `data.frame`s.
+
+This means that not only users will be able to get started with your package faster because of this familiarity, but you will also immediately benefit for the huge ecosystem of functions and packages working on data.frames, such as the [tidyverse](https://tidyverse.org). If you want some examples, this is what we did in the [linelist](https://github.com/epiverse-trace/linelist), [pavo](https://github.com/rmaia/pavo), [scoringutils](https://github.com/epiforecasts/scoringutils), and [vaccineff](https://github.com/epiverse-trace/vaccineff) R packages.
```

Comment:
```suggestion
This means that not only users will be able to get started with your package faster because of this familiarity, but you will also immediately benefit for the huge ecosystem of functions and packages working on `data.frame`s, such as the [tidyverse](https://tidyverse.org). If you want some examples, this is what we did in the [linelist](https://github.com/epiverse-trace/linelist), [pavo](https://github.com/rmaia/pavo), [scoringutils](https://github.com/epiforecasts/scoringutils), and [vaccineff](https://github.com/epiverse-trace/vaccineff) R packages.
```